framework/src/modules/bft/api.ts

```diff
@@ -12,45 +12,253 @@
+import {
+	bftVotesSchema,
+	BFTVotes,
+	BFTParameters,
+	ValidatorsHashValidator,
+	ValidatorsHashInput,
+	validatorsHashInputSchema,
+	bftParametersSchema,
+	BFTVotesActiveValidatorsVoteInfo,
+} from './schemas';
+import { BFTHeights, Validator, ValidatorsAPI } from './types';
+import { BFTParameterNotFoundError } from './errors';
 
 export interface BlockHeaderAsset {
 	maxHeightPrevoted: number;
 	maxHeightPreviouslyForged: number;
 }
 
 export class BFTAPI extends BaseAPI {
-	public async verifyGeneratorInfo(
-		_apiContext: APIContext,
-		_generatorStore: GeneratorStore,
-		_info: {
-			address: Buffer;
-			height: number;
-			maxHeightPrevoted: number;
-			maxHeightPreviouslyForged: number;
-			override?: boolean;
-		},
-		// eslint-disable-next-line @typescript-eslint/no-empty-function
-	): Promise<void> {}
-
-	// eslint-disable-next-line @typescript-eslint/require-await
-	public async getValidators(_apiContext: ImmutableAPIContext): Promise<Validator[]> {
-		return [];
-	}
-
-	// eslint-disable-next-line @typescript-eslint/require-await
-	public async getBFTHeights(_apiClient: ImmutableAPIContext): Promise<BFTVotes> {
+	private _validatorsAPI!: ValidatorsAPI;
+
+	public addDependencies(validatorsAPI: ValidatorsAPI): void {
+		this._validatorsAPI = validatorsAPI;
+	}
+
+	public areHeadersContradicting(
+		bftHeader1: ChainBlockHeader,
+		bftHeader2: ChainBlockHeader,
+	): boolean {
+		if (bftHeader1.id.equals(bftHeader2.id)) {
+			return false;
+		}
+		return areDistinctHeadersContradicting(bftHeader1, bftHeader2);
+	}
+
+	public async isHeaderContradictingChain(
+		context: ImmutableAPIContext,
+		header: BlockHeader,
+	): Promise<boolean> {
+		const votesStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_VOTES);
+		const bftVotes = await votesStore.getWithSchema<BFTVotes>(EMPTY_KEY, bftVotesSchema);
+		for (const bftBlock of bftVotes.blockBFTInfos) {
+			if (bftBlock.generatorAddress.equals(header.generatorAddress)) {
+				return areDistinctHeadersContradicting(bftBlock, header);
+			}
+		}
+		return false;
+	}
+
+	public async existBFTParameters(context: ImmutableAPIContext, height: number): Promise<boolean> {
+		const paramsStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_PARAMETERS);
+		return paramsStore.has(intToBuffer(height, 4, BIG_ENDIAN));
+	}
+
+	public async getBFTParameters(
+		context: ImmutableAPIContext,
+		height: number,
+	): Promise<BFTParameters> {
+		const paramsStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_PARAMETERS);
+		return getBFTParameters(paramsStore, height);
+	}
+
+	public async getBFTHeights(context: ImmutableAPIContext): Promise<BFTHeights> {
+		const votesStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_VOTES);
+		const bftVotes = await votesStore.getWithSchema<BFTVotes>(EMPTY_KEY, bftVotesSchema);
 		return {
-			maxHeightPrecommited: 0,
-			maxHeightPrevoted: 0,
+			maxHeightPrevoted: bftVotes.maxHeightPrevoted,
+			maxHeightPrecommited: bftVotes.maxHeightPrecommited,
```

Comment:
`maxHeightPrecommited` => `maxHeightPrecommitted` is done in next PR https://github.com/LiskHQ/lisk-sdk/pull/6796